server.py

```diff
@@ -56,11 +78,16 @@ def file_upload(number):
     expected = expected.strip()
     uploaded = uploaded.strip()
     ans = (uploaded == expected)
-    usernames[u_name].append(Submission(question=number, time=time,
+    if not u_name in usernames.keys():
+    	usernames[u_name] = User()
```

Comment:
this pattern can be avoided if you use defaultdicts properly. If I remember correctly `usernames` is a default dict.